# Data and scripts for reproducing the analyses

This repository contains data and scripts used in our recent manuscript ([Chng *et al*, Nature Medicine, 2020](#citation)).

## Directory structure

 - [metadata](metadata): Metadata files mapping sequencing IDs to site information
 - [output_tables](output_tables): Data files generated by analysis scripts
 - [plots](plots): Figures used in the manuscript
 - [scripts](scripts): R markdown files (with rendered html) for the analysis
 - [tables](tables): Input data used for generating the figures
 
## Downloading this repository

1. Clone the core component using git
```sh
git clone https://github.com/csb5/hospital_microbiome.git
```

2. Download large datasets using [git lfs](https://git-lfs.github.com/)
```sh
cd hospital_microbiome && git lfs pull
```

## Raw data availability

 - Raw sequencing data has been uploaded to [European Nucleotide Archive](https://www.ebi.ac.uk/ena/data/view/PRJEB31632)
 - Assemblies and annotations are available at [Figshare](https://figshare.com/articles/Cartography_of_opportunistic_pathogens_and_antibiotic_resistance_genes_in_a_tertiary_hospital_environment/11683350)
 - Source code and software for visualizing the data on 3D models are available on [github](https://github.com/CSB5/hospital_microbiome_explorer)

## Citation

Chng *et al*. (2020). Cartography of opportunistic pathogens and antibiotic resistance genes in a tertiary hospital environment. Nature Medicine. *In press*.

